src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -35,15 +35,21 @@ internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock
                     // PROTOTYPE (https://github.com/dotnet/roslyn/issues/9819): Try to integrate instrumentation with lowering, to avoid an extra pass over the bound tree.
                     BoundBlock newMethodBody = InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, payloadField, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans);
 
-                    // Synthesize the initialization of the instrumentation payload array. It should be done either statically or with concurrency-safe code:
+                    // Synthesize the initialization of the instrumentation payload array, using concurrency-safe code:
                     //
                     // if (payloadField == null)
                     //     Instrumentation.CreatePayload(mvid, method, ref payloadField, payloadLength);
 
-                    // PROTOTYPE (https://github.com/dotnet/roslyn/issues/9812):
-                    // The containing module's mvid should be computed statically and stored in a static field rather than being
-                    // recomputed in each method prologue.
-                    BoundExpression mvid = factory.Property(factory.Property(factory.TypeofBeforeRewriting(method.ContainingType), "Module"), "ModuleVersionId");
+                    // Guarantee that PrivateImplementationDetails will initialize its MVID field.
+                    PrivateImplementationDetails privateImplementationClass = compilationState.ModuleBuilderOpt.GetPrivateImplClass(null, diagnostics);
+                    if (privateImplementationClass.StaticConstructor == null)
+                    {
+                        // This is concurrency safe because the StaticConstructor setter ignores all but the first assignment.
+                        privateImplementationClass.StaticConstructor = new SynthesizedPrivateImplementationStaticConstructor(compilationState.ModuleBuilderOpt.SourceModule, privateImplementationClass, factory);
```

Comment:
Actually, we can even postpone creation/addition of the constructor symbol till the "freeze" moment. At that time we should have complete list of initializations that constructor should perform.